### PR TITLE
Goodbye, Translog Views

### PR DESCRIPTION
**Pull request by bleskes**
During peer recoveries, we need to copy over lucene files and replay the operations they miss from the source translog. Guaranteeing that translog files are not cleaned up has seen many iterations overtime. Back in the old 1.0 days, recoveries went through the Engine and actively prevented both translog cleaning and lucene commits. We then moved to a notion called Translog Views, which allowed the recovery code to "acquire" a view into the translog which is then guaranteed to be kept around until the view is closed. The Engine code was free to commit lucene and do what it ever it wanted without coordinating with recoveries. Translog file deletion logic was based on reference counting on the file level. Those counters were incremented when a view was acquired but also when the view was used to create a `Snapshot` that allowed you to read operations from the files. At some point we removed the file based counting complexity in favor of constructs on the Translog level that just keep track of "open" views and the minimum translog generation they refer to. To do so, Views had to be kept around until the last snapshot that was made from them was consumed. This was fine in recovery code but lead to [a subtle bug](https://github.com/elastic/elasticsearch/pull/25862) in the [Primary Replica Resyncer](https://github.com/elastic/elasticsearch/pull/25862). 

Concurrently, we have developed the notion of a `TranslogDeletionPolicy` which is responsible for the liveness aspect of translog files. This class makes it very simple to take translog Snapshot into account for keep translog files around, allowing people that just need a snapshot to just take a snapshot and not worry about views and such. Recovery code which actually does need a view can now prevent trimming by acquiring a simple retention lock (a `Closable`). This removes the need for the notion of a View.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/translog/TranslogDeletionPolicy.java [hunk 1]
```diff
@@ -66,22 +85,42 @@ public synchronized void setRetentionAgeInMillis(long ageInMillis) {
     }
 
     /**
-     * acquires the basis generation for a new view. Any translog generation above, and including, the returned generation
-     * will not be deleted until a corresponding call to {@link #releaseTranslogGenView(long)} is called.
+     * acquires the basis generation for a new view or snapshot. Any translog generation above, and including, the returned generation
+     * will not be deleted until the returned {@link Releasable} is closed.
      */
-    synchronized void acquireTranslogGenForView(final long genForView) {
+    synchronized Releasable acquireTranslogGen(final long genForView) {
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/index/translog/TranslogDeletionPolicy.java]: Change the parameter name to not refer to views?

File: core/src/main/java/org/elasticsearch/index/translog/TranslogDeletionPolicy.java [hunk 2]
```diff
@@ -66,22 +85,42 @@ public synchronized void setRetentionAgeInMillis(long ageInMillis) {
...
+    synchronized int pendingTranslogRefCount() {
         return translogRefCounts.size();
     }
 
     /**
-     * releases a generation that was acquired by {@link #acquireTranslogGenForView(long)}
+     * releases a generation that was acquired by {@link #acquireTranslogGen(long)}
      */
-    synchronized void releaseTranslogGenView(long translogGen) {
+    private synchronized void releaseTranslogGenView(long translogGen) {
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/index/translog/TranslogDeletionPolicy.java]: Change the method name to remove the mention of views?

File: core/src/main/java/org/elasticsearch/index/translog/TranslogDeletionPolicy.java [hunk 3]
```diff
@@ -20,14 +20,28 @@
...
+    public void assertNoOpenTranslogRefs() {
+        if (openTranslogRef.isEmpty() == false) {
+            AssertionError e = new AssertionError("not all translog generations have been released");
+            openTranslogRef.values().forEach(e::addSuppressed);
+            throw e;
+        }
+    }
+
     /**
      * Records how many views are held against each
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/index/translog/TranslogDeletionPolicy.java]: Remove mention of views?

File: core/src/main/java/org/elasticsearch/index/translog/TranslogDeletionPolicy.java [hunk 4]
```diff
@@ -99,7 +138,7 @@ synchronized void releaseTranslogGenView(long translogGen) {
      * @param writer  current translog writer
      */
     synchronized long minTranslogGenRequired(List<TranslogReader> readers, TranslogWriter writer) throws IOException {
-        long minByView = getMinTranslogGenRequiredByViews();
+        long minByView = getMinTranslogGenRequiredByLocks();
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/index/translog/TranslogDeletionPolicy.java]: Remove mention of views?

File: core/src/main/java/org/elasticsearch/index/translog/Translog.java [hunk 5]
```diff
@@ -1627,9 +1595,9 @@ void deleteReaderFiles(TranslogReader reader) {
             reader.path().resolveSibling(getCommitCheckpointFileName(reader.getGeneration())));
     }
 
-    void closeFilesIfNoPendingViews() throws IOException {
+    void closeFilesIfNoPendingRetentionLocks() throws IOException {
         try (ReleasableLock ignored = writeLock.acquire()) {
-            if (closed.get() && deletionPolicy.pendingViewsCount() == 0) {
+            if (closed.get() && deletionPolicy.pendingTranslogRefCount() == 0) {
                 logger.trace("closing files. translog is closed and there are no pending views");
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/index/translog/Translog.java]: Remove mention of views from this trace message: `closing files; translog is closed and there are no pending retention locks`

File: core/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java [hunk 6]
```diff
@@ -249,7 +249,7 @@ public ClusterState execute(ClusterState currentState) {
                 }
                 SnapshotsInProgress snapshots = currentState.custom(SnapshotsInProgress.TYPE);
                 if (snapshots == null || snapshots.entries().isEmpty()) {
-                    // Store newSnapshot here to be processed in clusterStateProcessed
+                    // Store newSnapshotFromGen here to be processed in clusterStateProcessed
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java]: I don't think so. 😄

File: core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java [hunk 7]
```diff
@@ -177,15 +180,15 @@ public RecoveryResponse recoverToTarget() throws IOException {
             runUnderPrimaryPermit(() -> shard.initiateTracking(request.targetAllocationId()));
 
             try {
-                prepareTargetForTranslog(translogView.estimateTotalOperations(startingSeqNo));
+                prepareTargetForTranslog(translog.estimateTotalOperationsFromMinSeq(startingSeqNo));
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: Let's avoid invoking `Translog#estimateTotalOperationsFromMinSeq` twice here?

File: core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java [hunk 8]
```diff
@@ -234,11 +236,12 @@ boolean isTranslogReadyForSequenceNumberBasedRecovery(final Translog.View transl
             logger.trace("all operations up to [{}] completed, checking translog content", endingSeqNo);
 
             final LocalCheckpointTracker tracker = new LocalCheckpointTracker(shard.indexSettings(), startingSeqNo, startingSeqNo - 1);
-            final Translog.Snapshot snapshot = translogView.snapshot(startingSeqNo);
-            Translog.Operation operation;
-            while ((operation = snapshot.next()) != null) {
-                if (operation.seqNo() != SequenceNumbersService.UNASSIGNED_SEQ_NO) {
-                    tracker.markSeqNoAsCompleted(operation.seqNo());
+            try(Translog.Snapshot snapshot = shard.getTranslog().newSnapshotFromMinSeqNo(startingSeqNo)) {
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: Nit: space between `try` and `(`.

**Review by jasontedor — COMMENTED** [review]
In general it looks good. I think that main point of contention is abstraction for the chained listeners in the primary/replica resync code. I think it's unnecessary and makes the code hard to follow, how it was (and what I propose) is clearer. Also, I left a comment about the `InternalEngineTests` although I'm not sure if anything can be done (I leave that to you to figure out). The rest I think are minor comments.

File: core/src/main/java/org/elasticsearch/index/shard/PrimaryReplicaSyncer.java [hunk 9]
```diff
@@ -79,31 +81,16 @@ void setChunkSize(ByteSizeValue chunkSize) { // only settable for tests
...
-                    e.addSuppressed(inner);
-                }
-                listener.onFailure(e);
-            }
-        };
         try {
             final long startingSeqNo = indexShard.getGlobalCheckpoint() + 1;
-            Translog.Snapshot snapshot = view.snapshot(startingSeqNo);
+            Translog.Snapshot snapshot = indexShard.getTranslog().newSnapshotFromMinSeqNo(startingSeqNo);
+            listener = chain(wrap(r -> snapshot.close(), e -> {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/shard/PrimaryReplicaSyncer.java]: These things are subjecting. I personally prefer the wrapping because it allows to ignore the crud of wrapping and focus on the functionality. Same goes for not re-assigning the listener (I prefer my version as it the complexities are dealt with in the same place, when we wrap). I don't feel strongly about it and will happily go along with your version.

File: core/src/main/java/org/elasticsearch/index/shard/PrimaryReplicaSyncer.java [hunk 10]
```diff
@@ -112,6 +99,11 @@ public void onFailure(Exception e) {
             Translog.Snapshot wrappedSnapshot = new Translog.Snapshot() {
 
                 @Override
+                public void close() throws IOException {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/shard/PrimaryReplicaSyncer.java]: sure

File: core/src/main/java/org/elasticsearch/index/translog/Translog.java [hunk 11]
```diff
@@ -574,23 +575,53 @@ public long getLastSyncedGlobalCheckpoint() {
...
+            return newMultiSnapshot(snapshots);
+        }
+    }
+
+    private Snapshot newMultiSnapshot(TranslogSnapshot[] snapshots) throws IOException {
+        final Closeable onClose;
+        if (snapshots.length == 0) {
+            onClose = () -> {};
+        } else {
+            assert Arrays.stream(snapshots).map(BaseTranslogReader::getGeneration).min(Long::compareTo).get()
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/translog/Translog.java]: I don't think so? acquiring the min gen from the translog deletion policy will keep all the other ones around which is what we care about here?

File: core/src/main/java/org/elasticsearch/index/translog/Translog.java [hunk 12]
```diff
@@ -574,23 +575,53 @@ public long getLastSyncedGlobalCheckpoint() {
...
+                == snapshots[0].generation : "first reader generation of " + snapshots[0].generation + " is not the smallest";
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/translog/Translog.java]: ++

File: core/src/main/java/org/elasticsearch/index/translog/Translog.java [hunk 13]
```diff
@@ -605,30 +636,27 @@ public Snapshot newSnapshot(long minGeneration) {
...
     /**
-     * Returns a view into the current translog that is guaranteed to retain all current operations
-     * while receiving future ones as well
+     * Acquires a lock on the translog files, preventing them from being trimmed
      */
-    public Translog.View newView() {
+    public Closeable acquireRetentionLock() {
         try (ReleasableLock lock = readLock.acquire()) {
             ensureOpen();
             final long viewGen = getMinFileGeneration();
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/translog/Translog.java]: yeah. That works. Will add.

File: core/src/main/java/org/elasticsearch/index/translog/Translog.java (hunk 5 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/translog/Translog.java]: good catch.

File: core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java [hunk 14]
```diff
@@ -140,11 +142,13 @@ public RecoveryResponse recoverToTarget() throws IOException {
             assert targetShardRouting.initializing() : "expected recovery target to be initializing but was " + targetShardRouting;
         });
 
-        try (Translog.View translogView = shard.acquireTranslogView()) {
+        try (Closeable translogRetentionLock = shard.acquireTranslogRetentionLock()) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: sure.

File: core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java (hunk 7 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: This could have changed? prepare target has a network call in it?

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 15]
```diff
@@ -1555,10 +1556,10 @@ public void onSettingsChanged() {
         }
     }
 
-    public Translog.View acquireTranslogView() {
+    public Closeable acquireTranslogRetentionLock() {
         Engine engine = getEngine();
         assert engine.getTranslog() != null : "translog must not be null";
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: yes.

File: core/src/main/java/org/elasticsearch/index/translog/TranslogDeletionPolicy.java (hunk 3 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/translog/TranslogDeletionPolicy.java]: good catch.

File: core/src/main/java/org/elasticsearch/index/translog/TranslogDeletionPolicy.java [hunk 16]
```diff
@@ -66,22 +85,42 @@ public synchronized void setRetentionAgeInMillis(long ageInMillis) {
     }
 
     /**
-     * acquires the basis generation for a new view. Any translog generation above, and including, the returned generation
-     * will not be deleted until a corresponding call to {@link #releaseTranslogGenView(long)} is called.
+     * acquires the basis generation for a new view or snapshot. Any translog generation above, and including, the returned generation
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/translog/TranslogDeletionPolicy.java]: ++

File: core/src/main/java/org/elasticsearch/index/translog/TranslogDeletionPolicy.java (hunk 2 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/translog/TranslogDeletionPolicy.java]: ++

File: core/src/main/java/org/elasticsearch/index/translog/TranslogDeletionPolicy.java (hunk 1 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/translog/TranslogDeletionPolicy.java]: ++

File: core/src/main/java/org/elasticsearch/index/translog/TranslogDeletionPolicy.java (hunk 4 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/translog/TranslogDeletionPolicy.java]: ++

File: core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java [hunk 17]
```diff
@@ -280,6 +280,12 @@ public EngineConfig copy(EngineConfig config, EngineConfig.OpenMode openMode, An
     @After
     public void tearDown() throws Exception {
         super.tearDown();
+        if (engine != null && engine.isClosed.get() == false) {
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java]: yeah, I was going back and forth between static maps in the assertion and how it's done now. The problem with static maps is that you need to clean them and it becomes ugly. I followed the same pattern we have in other places. The good news is that we also have file system leakage check and when people forget to close a snapshot it trips.

File: core/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java (hunk 6 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java]: :(

**Comment by bleskes** [comment]
@jasontedor thx. I addressed all your feedback. Can you take another look?

File: core/src/main/java/org/elasticsearch/index/shard/PrimaryReplicaSyncer.java (hunk 9 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/index/shard/PrimaryReplicaSyncer.java]: I disagree it allows focusing on the functionality, to chase down what is really being executed here you have to run off and grok two methods, it's really not straightforward at all. With the listener defined front and center you can immediately see what is happening, no need to chase anything down. Sorry, I feel very strongly about this one.

**Review by jasontedor — APPROVED** [review]
LGTM.

**Comment by bleskes** [comment]
Thanks @jasontedor
